agent/src/main/java/com/code_intelligence/jazzer/agent/CoverageIdStrategy.kt

```diff
@@ -0,0 +1,189 @@
+     * The caller *must* also call [commitIdCount] once it has instrumented that class, even if instrumentation fails.
+     */
+    @Throws(CoverageIdException::class)
+    fun obtainFirstId(className: String): Int
+
+    /**
+     * Records the number of coverage IDs used to instrument the class specified in a previous call to [obtainFirstId].
+     * If instrumenting the class should fail, this function must still be called. In this case, [idCount] is set to 0.
+     */
+    @Throws(CoverageIdException::class)
+    fun commitIdCount(idCount: Int)
+}
+
+/**
+ * An unsynchronized strategy for coverage ID generation that simply increments a global counter.
+ */
+internal class TrivialCoverageIdStrategy : CoverageIdStrategy {
+    private var nextEdgeId = 0
+
+    override fun obtainFirstId(className: String) = nextEdgeId
+
+    override fun commitIdCount(idCount: Int) {
+        nextEdgeId += idCount
+    }
+}
+
+/**
+ * Reads the [FileChannel] to the end as a UTF-8 string.
+ */
+private fun FileChannel.readFully(): String {
+    check(size() <= Int.MAX_VALUE)
+    val buffer = ByteBuffer.allocate(size().toInt())
+    while (buffer.hasRemaining()) {
+        when (read(buffer)) {
+            0 -> throw IllegalStateException("No bytes read")
+            -1 -> break
+        }
+    }
+    return String(buffer.array())
+}
+
+/**
+ * Appends [string] to the end of the [FileChannel].
+ */
+private fun FileChannel.append(string: String) {
+    position(size())
+    write(ByteBuffer.wrap(string.toByteArray()))
+}
+
+/**
+ * A strategy for coverage ID generation that synchronizes the IDs assigned to a class with other processes via the
+ * specified [idSyncFile].
+ * This class takes care of synchronizing the access to the file between multiple processes as long as the general
+ * contract of [CoverageIdStrategy] is followed.
+ */
+internal class SynchronizedCoverageIdStrategy(private val idSyncFile: Path) : CoverageIdStrategy {
+    var idFileLock: FileLock? = null
+
+    var cachedFirstId: Int? = null
+    var cachedClassName: String? = null
+    var cachedIdCount: Int? = null
+
+    /**
+     * Obtains a coverage ID for [className] such that all cooperating agent processes will obtain the same ID.
+     * There are two cases to consider:
+     * - This agent process is the first to encounter [className], i.e., it does not find a record for that class in
+     *   [idSyncFile]. In this case, a lock on the file is held until the class has been instrumented and a record with
+     *   the required number of coverage IDs has been added.
+     * - Another agent process has already encountered [className], i.e., there is a record that class in [idSyncFile].
+     *   In this case, the lock on the file is returned immediately and the extracted first coverage ID is returned to
+     *   the caller. The caller is still expected to call [commitIdCount] so that desynchronization can be detected.
```

Comment:
This is stable since JaCoCo visits language constructs in the order they are defined in the bytecode. JaCoCo also depends on this being the case, so I think we are safe here.

Reply:
Perfect!